engine/execution/ingestion/stop_control.go

```diff
@@ -0,0 +1,221 @@
+package ingestion
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// StopControl is a specialized component used by ingestion.Engine to encapsulate
+// control of pausing/stopping blocks execution.
+// It is intended to work tightly with the Engine, not as a general mechanism or interface.
+// StopControl follows states described in StopState
+type StopControl struct {
+	sync.RWMutex
+	height             uint64
+	crash              bool
+	stopAfterExecuting flow.Identifier
+	log                zerolog.Logger
+	state              StopControlState
+	lastExecutedHeight uint64 // last executed height, conveniently defaults to zero
+}
+
+type StopControlState byte
+
+const (
+	// StopControlOff default state, envisioned to be used most of the time. Stopping module is simply off,
+	// blocks will be processed "as usual".
+	StopControlOff StopControlState = iota
+
+	// StopControlSet means stop height is set but not reached yet, and nothing related to stopping happened yet.
```

Comment:
I found the word `reached` a bit unclear. 

Also please refer to my comments to add comments to the state variables, such as `height`, `crash`, etc